flink-python/pyflink/datastream/window.py

```diff
@@ -517,3 +538,721 @@ def __init__(self,
+    """
+    A WindowAssigner that windows elements into sliding windows based on the count number of
+    the elements. Windows can possibly overlap.
+    """
+
+    def __init__(self, window_size: int, window_slide: int):
+        """
+        Windows this KeyedStream into sliding count windows.
+        :param window_size: The size of the windows in number of elements.
+        :param window_slide: The slide interval in number of elements.
+        """
+        self._window_size = window_size
+        self._window_slide = window_slide
+        self._count = None  # type: ValueState
+        self._counter_state_descriptor = ReducingStateDescriptor(
+            "slide-count-assigner", lambda a, b: a + b, Types.LONG())
+
+    def assign_windows(self,
+                       element: T,
+                       timestamp: int,
+                       context: 'WindowAssigner.WindowAssignerContext') -> Collection[W]:
+        count_descriptor = ValueStateDescriptor('slide-count-assigner', Types.LONG())
+        self._count = context.get_runtime_context().get_state(count_descriptor)
+        count_value = self._count.value()
+        if count_value is None:
+            current_count = 0
+        else:
+            current_count = count_value
+        self._count.update(current_count + 1)
+        last_id = current_count // self._window_slide
+        last_start = last_id * self._window_slide
+        last_end = last_start + self._window_size - 1
+        windows = []
+        while last_id >= 0 and last_start <= current_count <= last_end:
+            if last_start <= current_count <= last_end:
+                windows.append(CountWindow(last_id))
+            last_id -= 1
+            last_start -= self._window_slide
+            last_end -= self._window_slide
+        return windows
+
+    def get_default_trigger(self, env) -> Trigger[T, W]:
+        return CountTrigger(self._window_size)
+
+    def get_window_serializer(self) -> TypeSerializer[W]:
+        return CountWindowSerializer()
+
+    def is_event_time(self) -> bool:
+        return False
+
+    def __repr__(self):
+        return "CountSlidingWindowAssigner(%s, %s)" % (self._window_size, self._window_slide)
+
+
+class SessionWindowAssigner(MergingWindowAssigner[T, TimeWindow]):
+    """
+        WindowAssigner that windows elements into sessions based on the timestamp. Windows cannot
+        overlap.
+        """
+
+    def __init__(self, session_gap: int, is_event_time: bool):
+        if session_gap <= 0:
+            raise Exception("SessionWindowAssigner parameters must satisfy 0 < size")
+
+        self._session_gap = session_gap
+        self._is_event_time = is_event_time
+
+    def merge_windows(self,
+                      windows: Iterable[W],
+                      callback: 'MergingWindowAssigner.MergeCallback[W]') -> None:
+        window_list = [w for w in windows]
+        window_list.sort()
+        for i in range(1, len(window_list)):
```

Comment:
This logic is incorrect when receiving late records, checkout java implementation